resolver/hosts_file_resolver.go

```diff
@@ -144,27 +151,76 @@ func (r *HostsFileResolver) resolve(req *dns.Msg, question dns.Question, domain
 	return response
 }
 
-func (r *HostsFileResolver) parseHostsFile(ctx context.Context) error {
-	const maxErrorsPerFile = 5
-
-	if r.cfg.Filepath == "" {
+func (r *HostsFileResolver) loadSources(ctx context.Context) error {
+	if !r.IsEnabled() {
 		return nil
 	}
 
-	f, err := os.Open(r.cfg.Filepath)
+	//nolint:ineffassign,staticcheck // keep `ctx :=` so if we use ctx in the future, we use the correct one
+	consumersGrp, ctx := jobgroup.WithContext(ctx)
+	defer consumersGrp.Close()
+
+	producersGrp := jobgroup.WithMaxConcurrency(consumersGrp, r.cfg.Loading.Concurrency)
+	defer producersGrp.Close()
+
+	producers := parcour.NewProducersWithBuffer[*HostsFileEntry](producersGrp, consumersGrp, producersBuffCap)
+	defer producers.Close()
```

Comment:
Basically a `JobGroup` is a scope for goroutines: the `defer grp.Close()` ensures no goroutine continues running after the function returns. Groups can be nested: here `producersGrp` is a child of `consumersGrp`.
These are ideas from [structured concurrency](https://vorpus.org/blog/notes-on-structured-concurrency-or-go-statement-considered-harmful/) (highly recommend that post, it's one of the most influential CS articles I've read).

Having a clear scope for goroutines also allows for clear failure propagation:
- when explicitly checking for goroutine errors with `Wait`:
  - if any goroutines returned errors, it returns them (just like `errgroup.Group`)
  - if any goroutines panicked, it propagates the panic to the current goroutine
- when we leave the function without explicitly waiting for goroutine, `Close` handles the failures:
  - if any goroutines returned errors, it propagates them to the parent `JobGroup` (or panics if there is no parent)
  - if any goroutines panicked, it propagates the panic to the current goroutine

`producers` is an abstraction based on `JobGroup`s to make the pattern we have here of producing host file entries from sources in parallel and consuming them from a single goroutine simple.

The fact that `producers` uses the `JobGroup`s we created means we know for sure it won't leave anything running since we're closing them before returning, and we can customize how those goroutines run.